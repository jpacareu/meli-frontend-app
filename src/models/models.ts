import { History } from 'history';

export interface IState {
	items: IItem[];
	searchBox: string;
	categories: string[];
	loadingSearch: boolean;
	detailProduct: {
		item: IDetailItem;
	} | null;
}

export interface IActions {
	search: (...rest: any) => {};
	getProductDetail: (...rest: any) => {};
	loading: (...rest: any) => {};
	received: (...rest: any) => {};
	setState: (...rest: any) => {};
	handleError: (...rest: any) => {};
}

export interface IStateHistory extends IState {
	history: History;
}

export interface IDetailItem {
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
	sold_quantity: number;
	description: string;
}

export interface IItem {
	id: string;
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
}

export type ICondition = 'new' | 'used';

export interface IPrice {
	currency: string;
	amount: number;
	decimals: number;
}

export interface IStore {
	Provider: any;
	connect: () => any;
	actions: IActions;
}
