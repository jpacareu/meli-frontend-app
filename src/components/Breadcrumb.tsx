import React from 'react';

interface IProps {
	list: string[];
}

export const Breadcrumb = (props: IProps) => {
	const list: string[] = props.list || [];
	return (
		<nav aria-label="breadcrumb">
			<ol className="breadcrumb-category d-md-flex">
				{list.map((el, i, arr) => {
					const isStrong = i === arr.length - 1 ? '__highlight' : '';
					return (
						<li
							key={i}
							className={`breadcrumb-category__item${isStrong}`}
							aria-current="page">
							{el}
						</li>
					);
				})}
			</ol>
		</nav>
	);
};
