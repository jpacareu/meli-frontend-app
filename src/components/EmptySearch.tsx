export const EmptySearch = () => {
	return null;
};
