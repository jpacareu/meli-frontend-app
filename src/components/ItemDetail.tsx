import React from 'react';
import { IDetailItem, ICondition, IPrice } from '../models/models';
import { formatAmount, formatCurrency } from '../utils/money';

interface IProps {
	detail: IDetailItem;
}

export const ItemDetail = (props: IProps) => {
	const { detail } = props;

	return (
		<div className="item-detail row">
			<div className="col-md-7">
				<Image title={detail.title} picture={detail.picture} />
			</div>
			<div className="col-md-4 col-sm-12">
				<section className="item-detail__main-info">
					<div className="item-detail__condition-and-sold d-flex">
						<Condition condition={detail.condition} />
						<span> - </span>
						<Sold sold={detail.sold_quantity} />
					</div>
					<Title title={detail.title} />
					<Price price={detail.price} />
					<ButtonComprar />
				</section>
			</div>
			<div className="col-md-7 col-sm-12">
				<Detalle description={detail.description} />
			</div>
		</div>
	);
};

export const Image = (props: { title: string; picture: string }) => {
	return (
		<img className="item-detail__image" alt={props.title} src={props.picture} />
	);
};
export const Detalle = (props: { description: string }) => {
	return (
		<section>
			<summary className="item-detail__descripcion-label">
				Descripci√≥n del producto
			</summary>
			<div className="item-detail__descripcion-text">{props.description}</div>
		</section>
	);
};

export const Condition = (props: { condition: ICondition }) => {
	return (
		<div className="item-detail__condition">
			{props.condition === 'new' ? 'Nuevo' : 'Usado'}
		</div>
	);
};

export const Sold = (props: { sold: number }) => {
	return <div className="item-detail__sold">{props.sold}</div>;
};

export const Title = (props: { title: string }) => {
	return <div className="item-detail__title">{props.title}</div>;
};

export const Price = (props: { price: IPrice }) => {
	return (
		<div className="item-detail__price">{`${formatCurrency(
			props.price.currency
		)} ${formatAmount(props.price.amount)}`}</div>
	);
};

export const ButtonComprar = () => {
	return <button className="item-detail__button">Comprar</button>;
};
