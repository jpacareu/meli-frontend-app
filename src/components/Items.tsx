import React from 'react';
import { IStateHistory, IItem, IPrice } from '../models/models';
import freeShippingLogo from '../assets/ic_shipping.png';
import { formatAmount, formatCurrency } from '../utils/money';
type PushFunction = (location: { pathname: string }) => void;

interface IItemWithRouter {
	push: PushFunction;
	model: IItem;
}

export const Items = (props: IStateHistory) => {
	return (
		<div className="result-search">
			{props.items.map((elem, index) => (
				<Item key={index} push={props.history.push} model={elem} />
			))}
		</div>
	);
};

const goProductDetail = (id: string, push: PushFunction) => {
	return () =>
		push({
			pathname: `/items/${id}`
		});
};

export const Item = (props: IItemWithRouter) => {
	const { model, push } = props;

	return (
		<div className="col-md-12">
			<div className="item">
				<Image model={model} push={push} />
				<div>
					<div className="d-flex flex-row">
						<div className="item__price d-flex align-items-center">
							<Price price={model.price} />
							{model.free_shipping && <FreeShipping />}
						</div>
					</div>
					<Title title={model.title} />
				</div>
			</div>
		</div>
	);
};

const Title = (props: { title: string }) => (
	<p className="item__title">{props.title}</p>
);

export const FreeShipping = () => {
	return (
		<img
			className="item__free_shipping"
			alt="Free shipping"
			src={freeShippingLogo}
		/>
	);
};

const Image = (props: { model: IItem; push: PushFunction }) => {
	return (
		<img
			alt={props.model.title}
			className="item__image"
			onClick={goProductDetail(props.model.id, props.push)}
			src={props.model.picture}
		/>
	);
};

const Price = (props: { price: IPrice }) => {
	return (
		<div>{`${formatCurrency(props.price.currency)} ${formatAmount(
			props.price.amount
		)}`}</div>
	);
};
