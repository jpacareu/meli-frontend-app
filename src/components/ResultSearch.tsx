import React, { Fragment } from 'react';
import { connect } from '../store';
import Spinner from './Spinner';
import { IState, IStateHistory } from '../models/models';
import { Items } from './Items';
import { Breadcrumb } from './Breadcrumb';

export let ResultSearch = (props: IStateHistory) => {
	return props.loadingSearch ? (
		<Spinner />
	) : (
		<Fragment>
			<Breadcrumb list={props.categories} />
			<Items {...props} />
		</Fragment>
	);
};

export default ResultSearch = connect((state: IState) => ({ ...state }))(
	ResultSearch
);
