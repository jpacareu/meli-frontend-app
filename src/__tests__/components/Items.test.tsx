import React from 'react';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { IStateHistory } from '../../models/models';
import { Items, Item, FreeShipping } from '../../components/Items';
import { History } from 'history';

const props: IStateHistory = {
	items: [
		{
			id: '1',
			title: 'Title 1',
			condition: 'new',
			picture: 'http://mywebsite.com/image1.jpg',
			price: {
				amount: 1234,
				currency: 'USD',
				decimals: 0
			},
			free_shipping: true
		},
		{
			id: '2',
			title: 'Title 2',
			condition: 'new',
			picture: 'http://mywebsite.com/image2.jpg',
			price: {
				amount: 5678,
				currency: 'ARS',
				decimals: 0
			},
			free_shipping: false
		},
		{
			id: '3',
			title: 'Title 3',
			condition: 'new',
			picture: 'http://mywebsite.com/image3.jpg',
			price: {
				amount: 9999,
				currency: 'ARS',
				decimals: 0
			},
			free_shipping: false
		}
	],
	history: {
		push: jest.fn() as any
	} as History
} as IStateHistory;

let component: JSX.Element | null = null;
let tree: ReactTestRenderer;

describe('<Items />', () => {
	beforeEach(() => {
		component = <Items {...props} />;
		tree = renderer.create(component);
	});
	it('should render without crashing', () => {
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it("should render 3 'Item'", () => {
		const items = tree.root.findAllByType(Item);
		expect(items.length).toBe(3);
	});

	it("should call 'push' from history", () => {
		const image = tree.root.findAllByProps({ className: 'item__image' })[0];

		image.props.onClick();

		expect(props.history.push).toHaveBeenCalled();
	});

	it("should render 1 'FreeShipping' component", () => {
		component = <Items {...props} />;
		tree = renderer.create(component);

		const items = tree.root.findAllByType(FreeShipping);
		expect(items).toHaveLength(1);
	});
});
