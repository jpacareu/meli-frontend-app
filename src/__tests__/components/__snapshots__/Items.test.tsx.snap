// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Items /> should render without crashing 1`] = `
<div
  className="result-search"
>
  <div
    className="col-md-12"
  >
    <div
      className="item"
    >
      <img
        alt="Title 1"
        className="item__image"
        onClick={[Function]}
        src="http://mywebsite.com/image1.jpg"
      />
      <div>
        <div
          className="d-flex flex-row"
        >
          <div
            className="item__price d-flex align-items-center"
          >
            <div>
              U$S 1.234
            </div>
            <img
              alt="Free shipping"
              className="item__free_shipping"
              src="ic_shipping.png"
            />
          </div>
        </div>
        <p
          className="item__title"
        >
          Title 1
        </p>
      </div>
    </div>
  </div>
  <div
    className="col-md-12"
  >
    <div
      className="item"
    >
      <img
        alt="Title 2"
        className="item__image"
        onClick={[Function]}
        src="http://mywebsite.com/image2.jpg"
      />
      <div>
        <div
          className="d-flex flex-row"
        >
          <div
            className="item__price d-flex align-items-center"
          >
            <div>
              $ 5.678
            </div>
          </div>
        </div>
        <p
          className="item__title"
        >
          Title 2
        </p>
      </div>
    </div>
  </div>
  <div
    className="col-md-12"
  >
    <div
      className="item"
    >
      <img
        alt="Title 3"
        className="item__image"
        onClick={[Function]}
        src="http://mywebsite.com/image3.jpg"
      />
      <div>
        <div
          className="d-flex flex-row"
        >
          <div
            className="item__price d-flex align-items-center"
          >
            <div>
              $ 9.999
            </div>
          </div>
        </div>
        <p
          className="item__title"
        >
          Title 3
        </p>
      </div>
    </div>
  </div>
</div>
`;
